// wxml-parse/lib/parse/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseNode = exports.TYPES = exports.NODE_TYPES = void 0;
var NODE_TYPES;
(function (NODE_TYPES) {
    NODE_TYPES[NODE_TYPES["ELEMENT"] = 1] = "ELEMENT";
    NODE_TYPES[NODE_TYPES["TEXT"] = 3] = "TEXT";
    NODE_TYPES[NODE_TYPES["COMMENT"] = 8] = "COMMENT";
    NODE_TYPES[NODE_TYPES["ROOT"] = 0] = "ROOT";
})(NODE_TYPES = exports.NODE_TYPES || (exports.NODE_TYPES = {}));
var TYPES;
(function (TYPES) {
    TYPES[TYPES["TEXT"] = 0] = "TEXT";
    TYPES[TYPES["COMMENT"] = 1] = "COMMENT";
    TYPES[TYPES["TAG_OPEN"] = 2] = "TAG_OPEN";
    TYPES[TYPES["TAG_CLOSE"] = 3] = "TAG_CLOSE";
    TYPES[TYPES["ATTRIBUTE_NAME"] = 4] = "ATTRIBUTE_NAME";
    TYPES[TYPES["ATTRIBUTE_VALUE"] = 5] = "ATTRIBUTE_VALUE";
    // start of mustache
    TYPES[TYPES["MUSTACHE_OPEN"] = 6] = "MUSTACHE_OPEN";
    TYPES[TYPES["MUSTACHE_CLOSE"] = 7] = "MUSTACHE_CLOSE";
    // end of mustache
})(TYPES = exports.TYPES || (exports.TYPES = {}));
class BaseNode {
    constructor(type) {
        this.type = type;
        this.parentNode = null;
    }
}
exports.BaseNode = BaseNode;
